// uu_appg01_binarystore/index.js
"use strict";

module.exports = {
  get DaoFactory() {
    return require("uu_appg01_datastore").DaoFactory;
  },
  get Lock() {
    return require("./src/libs/binarystore-lock");
  },
  get UuBinaryDao() {
    return require("./src/libs/dao");
  },
  get BinaryStoreError() {
    return require("./src/libs/dao_errors/binarystore-error");
  },
  get DuplicateKey() {
    return require("./src/libs/dao_errors/duplicate-key.js");
  },
  get ExpiredLock() {
    return require("./src/libs/dao_errors/expired-lock.js");
  },
  get InvalidRevision() {
    return require("./src/libs/dao_errors/invalid-revision.js");
  },
  get LockDoesNotExist() {
    return require("./src/libs/dao_errors/lock-does-not-exist.js");
  },
  get MissingRevision() {
    return require("./src/libs/dao_errors/missing-revision.js");
  },
  get NoiExceeced() {
    return require("./src/libs/dao_errors/noi-exceeded.js");
  },
  get ObjectLocked() {
    return require("./src/libs/dao_errors/object-locked.js");
  },
  get ObjectNotFound() {
    return require("./src/libs/dao_errors/object-not-found.js");
  },
  get SobExceeded() {
    return require("./src/libs/dao_errors/sob-exceeded.js");
  },
  get SoiExceeded() {
    return require("./src/libs/dao_errors/soi-exceeded.js");
  },
  get UnexpectedError() {
    return require("./src/libs/dao_errors/unexpected-error.js");
  }
};
